sprites.py

```diff
@@ -182,47 +198,57 @@ def __init__(self, sprites, x, y, image):
 
     def set_image(self, image, i=0, dx=0, dy=0):
         ''' Add an image to the sprite. '''
-        while len(self.images) < i + 1:
-            self.images.append(None)
+        while len(self.cached_surfaces) < i + 1:
+            self.cached_surfaces.append(None)
             self._dx.append(0)
             self._dy.append(0)
-        self.images[i] = image
         self._dx[i] = dx
         self._dy[i] = dy
-        if hasattr(self.images[i], 'get_width'):
-            w = self.images[i].get_width()
-            h = self.images[i].get_height()
+        if isinstance(image, GdkPixbuf.Pixbuf) or \
+           isinstance(image, cairo.ImageSurface):
+            w = image.get_width()
+            h = image.get_height()
         else:
-            w, h = self.images[i].get_size()
+            w, h = image.get_size()
         if i == 0:  # Always reset width and height when base image changes.
-            self.rect[2] = w + dx
-            self.rect[3] = h + dy
+            self.rect.width = w + dx
+            self.rect.height = h + dy
         else:
-            if w + dx > self.rect[2]:
-                self.rect[2] = w + dx
-            if h + dy > self.rect[3]:
-                self.rect[3] = h + dy
+            if w + dx > self.rect.width:
+                self.rect.width = w + dx
+            if h + dy > self.rect.height:
+                self.rect.height = h + dy
+        if isinstance(image, cairo.ImageSurface):
+            self.cached_surfaces[i] = image
+        else:  # Convert to Cairo surface
+            surface = cairo.ImageSurface(
+                cairo.FORMAT_ARGB32, self.rect.width, self.rect.height)
+            context = cairo.Context(surface)
+            Gdk.cairo_set_source_pixbuf(context, image, 0, 0)
+            context.rectangle(0, 0, self.rect.width, self.rect.height)
+            context.fill()
+            self.cached_surfaces[i] = surface
 
     def move(self, pos):
         ''' Move to new (x, y) position '''
         self.inval()
-        self.rect[0], self.rect[1] = int(pos[0]), int(pos[1])
+        self.rect.x, self.rect.y = int(pos[0]), int(pos[1])
         self.inval()
 
     def move_relative(self, pos):
         ''' Move to new (x+dx, y+dy) position '''
         self.inval()
-        self.rect[0] += int(pos[0])
-        self.rect[1] += int(pos[1])
+        self.rect.x += int(pos[0])
+        self.rect.y += int(pos[1])
         self.inval()
 
     def get_xy(self):
         ''' Return current (x, y) position '''
-        return (self.rect[0], self.rect[1])
+        return (self.rect.x, self.rect.y)
```

Comment:
Might be good to remove redundant parenthesis

Reply:
Yeah Thanks

Reply:
> Might be good to remove redundant parenthesis

`sprites.py` is maintained in `Turtle Art` activity, changing it here makes future update difficult; if you want a change it should be from there and not here.